python/ray/experimental/workflow/step_executor.py

```diff
@@ -145,15 +132,116 @@ def commit_step(store: workflow_storage.WorkflowStorage,
+    | Step Type                   | state | output |
+    +-----------------------------+-------+--------+
+    | Function Step               | Y     | N      |
+    +-----------------------------+-------+--------+
+    | Virtual Actor Step          | Y     | Y      |
+    +-----------------------------+-------+--------+
+    | Readonly Virtual Actor Step | N     | Y      |
+    +-----------------------------+-------+--------+
+    """
+
+    def __init__(self, step_type: StepType, state: Any, output: Any,
+                 catch_exceptions: bool):
+        self.step_type = step_type
+        self.state = state
+        self.output = output
+        self.catch_exceptions = catch_exceptions
+        self.exception = None
+
+    def warp_run(self, func: Callable, *args, **kwargs) -> "ExecutionResult":
+        """Wrap the function and execute it."""
+        if self.step_type == StepType.FUNCTION:
+            ret = func(*args, **kwargs)
+            return ExecutionResult(self.step_type, ret, None,
+                                   self.catch_exceptions)
+        elif self.step_type == StepType.ACTOR_METHOD:
+            state, output = func(*args, **kwargs)
+            return ExecutionResult(self.step_type, state, output,
+                                   self.catch_exceptions)
+        elif self.step_type == StepType.READONLY_ACTOR_METHOD:
+            output = func(*args, **kwargs)
+            return ExecutionResult(self.step_type, None, output,
+                                   self.catch_exceptions)
+        else:
+            raise ValueError(f"Unknown StepType '{self.step_type}'")
+
+    def commit(self, step_id: "StepID", outer_most_step_id: "StepID",
+               last_step_of_workflow: bool):
+        """Commit the output of the workflow"""
+        if self.step_type == StepType.READONLY_ACTOR_METHOD:
+            # Readonly virtual actor does not write the storage.
+            return
+        is_nested = isinstance(self.state, Workflow)
+        if self.step_type != StepType.FUNCTION and is_nested:
+            # TODO(suquark): Support returning a workflow inside
+            # a virtual actor.
+            raise TypeError("Only a workflow step function "
+                            "can return a workflow.")
+
+        store = workflow_storage.get_workflow_storage()
+        # Save workflow output
+        commit_step(store, step_id, self.state, outer_most_step_id)
+
+        # We MUST execute the workflow after saving the output.
+        if is_nested:
+            if self.step_type == StepType.FUNCTION:
+                # execute sub-workflow
+                self.state = execute_workflow(self.state, outer_most_step_id,
+                                              last_step_of_workflow)
+            else:
+                assert False, "This code path should not be reachable."
+        elif last_step_of_workflow:
+            # advance the progress of the workflow
+            store.advance_progress(step_id)
```

Comment:
I don't feel this function belongs to commit. It's a little bit wired to have execution logic in the commit function.